src/Common/Config/ConfigProcessor.h

```diff
@@ -114,6 +116,9 @@ class ConfigProcessor
     static bool isPreprocessedFile(const std::string & config_path);
 
 #if USE_SSL
+    /// Decrypt elements in config with specified encryption attributes
+    void decryptEncryptedElements(LoadedConfig & loaded_config, bool load_encryption_codecs = true, bool decrypt_encrypted_values = true);
```

Comment:
It is not clear to me in which circumstances `load_encryption_codecs` should be false? What bad things will happen if it is `false`, i.e. will the function which is supposed to decrypt something still be able to decrypt something?

Similarily, it seems strange to have a parameter `decrypt_encrypted_values` in a function `decryptEncryptedElements`? If the value is `false`, the function is not doing the very thing it is supposed to do.

Maybe
- renaming the parameters helps,
- or splitting the function into multiple functions,
- or adding better docs,
- or removing the two parameters altogether (and letting function `decryptEncryptedElements` figure out if it needs to load codecs or decrypt something).

Reply:
Maybe you like to mention in the docs (l. 119) that `decryptEncryptedElements` is usually called after `savePreprocessedConfig` such that encrypted elements do not land plaintext on disk.

Reply:
@mgorshkov will explain in details. But in short: we need to decrypt both: config.xml and users.xml
Codecs are only defines in config.xml. So `load_encryption_codecs == true` when loading or re-loading config.xml file (and not users.xml) file. As it's done in code: config.xml is loaded first, then users.xml is loaded.

Reply:
I think, splitting decryptEncryptedElements into 2 separate functions for keys loading and decryption itself, instead of introducing 2 bool parameters will make things clearer.
We'd like to load keys/decrypt config values on startup BaseDaemon::initialize call unless the keys are defined in ZK, because this raises an exception (explained earlier), that's one of the reasons why this PR was created.
Another one is users.xml decryption.
The keys must be loaded once and only once (that's how CompressionCodecEncrypted::Configuration::loadImpl is designed), so that's why all this stuff with boolean flags is here.